docs/source/util.py

```diff
@@ -0,0 +1,174 @@
+#
+# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
+#
+
+import argparse
+import os
+import shutil
+import subprocess
+from collections import namedtuple
+from typing import Iterable, Optional
+
+Class = namedtuple(
+    "Class",
+    ["module", "methods", "attributes"],
+    defaults=["snowflake.snowpark", [], []],
+)
+Module = namedtuple(
+    "Module", ["name", "attributes", "functions", "classes", "exceptions"]
+)
+mapping = {
+    "Methods": "methods",
+    "Attributes": "attributes",
+    "Module Attributes": "attributes",
+    "Functions": "functions",
+    "Classes": "classes",
+    "Exceptions": "exceptions",
+}
+
+
+def autogen_and_parse_for_info(module_name: str, class_name: Optional[str] = None):
+    if class_name:
+        res = Class(module_name, [], [])
+        name = f"{module_name}.{class_name}"
+    else:
+        res = Module(module_name, [], [], [], [])
+        name = module_name
+
+    if not os.path.exists("temp"):
+        os.mkdir("temp")
+
+    rst_content = f"""
+
+.. currentmodule:: snowflake.snowpark
+
+.. autosummary::
+    :members:
+    :inherited-members:
+    :toctree: api/
+
+    {name}
+
+"""
+    fname = f"temp/{name}_temp.rst"
+    with open(fname, "w") as fp:
+        fp.write(rst_content)
+
+    subprocess.run(["sphinx-autogen", fname, "-o", "output", "-t", "_templates"])
+
+    section = ""
+
+    with open(f"output/{name}.rst") as fp:
+        for line in fp:
+            line = line.strip()
+            if line.startswith(".. autosummary::"):
+                continue
+            if line.startswith(".. rubric:: "):
+                line = line[12:]
+                for title in mapping.keys():
+                    if line.startswith(title):
+                        section = mapping[title]
+                        break
+                continue
+            if not section:
+                continue
+            if line and not line.isspace():
+                getattr(res, section).append(line.rstrip())
+            elif getattr(res, section):  # End of a section
+                section = ""
+
+    return res
+
+
+def generate_section(section: str, content: str):
+    if content:
+        return f"""
+.. rubric:: {section}
+
+.. autosummary::
+    :toctree: api/
+
+    {content}
+
+"""
+    else:
+        return ""
+
+
+def generate(title: str, module: str, classes: Iterable[str]) -> str:
+    import itertools
+
+    if classes:
+        results = [autogen_and_parse_for_info(module, c) for c in classes]
+        names = "\n    =".join(classes)
```

Comment:
I wonder if we could make `\n` + 4 spaces a variable, so that we don't need to count it every time.